### ISSUE TITLE
[Feature] Generalize prefill Piecewise CUDA Graph through the Omni ModelRunner abstraction

### ISSUE THREAD

**Issue by AkazaAkane** [sgl-project/sglang-omni#1033]
## Motivation

SGLang main provides generic prefill CUDA-graph support through `PrefillCudaGraphRunner` (`python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py`) and `cuda_graph_config.prefill`, with two backends: `breakable` (BCG, the CUDA default) and `tc_piecewise` (Piecewise CUDA Graph). *(In the release SGLang-Omni currently pins, `sglang==0.5.12.post1`, the same machinery lives in `model_executor/piecewise_cuda_graph_runner.py` as `PiecewiseCudaGraphRunner` / `BreakableCudaGraphRunner`; this issue uses main's names.)*

SGLang-Omni's shared `ModelRunner` (`sglang_omni/model_runner/base.py`) wraps the SGLang TP worker. Prefill batches that fall through to the standard path:

```python
self.tp_worker.forward_batch_generation(forward_batch)
```

enter SGLang main's graph-aware `ModelRunner.forward` and can use the prefill graph runner.

However, most Omni models override `custom_prefill_forward(...)` to inject custom inputs. These implementations call the model directly, initialize attention metadata themselves, and hard-code the opt-out:

```python
model_runner.attn_backend.init_forward_metadata(forward_batch)
logits_output = model(
    input_ids=forward_batch.input_ids,
    positions=positions,
    forward_batch=forward_batch,
    input_embeds=input_embeds,
)
return GenerationBatchResult(logits_output=logits_output, can_run_cuda_graph=False)
```

Concretely: `thinker_model_runner.py` (L324), `ming_thinker_model_runner.py` (L251), `models/qwen3_omni/talker_model_runner.py` (L525), and `models/qwen3_tts/model_runner.py` (L270) all return `can_run_cuda_graph=False`, and the MOSS-TTS runners follow the same direct-call pattern. This bypasses upstream's graph dispatch entirely, so prefill PCG becomes a per-model problem even though the shared `ModelRunner` abstraction was built to centralize forward execution.

Typical custom inputs are: precomputed or projected `input_embeds`, multimodal embeddings, deepstack visual embeddings, and model-specific prefill buffers.

## Goal

Make prefill PCG generally available to compatible SGLang-Omni AR models through the shared `ModelRunner` abstraction, without a per-model CUDA-graph runner, by:

1. Reusing the existing Omni prefill hook split — `before_prefill` as the prepare-prefill phase, `custom_prefill_forward` as an explicit opt-out for model-owned forwards.
2. Reusing SGLang main's prefill graph infrastructure end-to-end: `PrefillCudaGraphRunner`, `PrefillInputBuffers`, `CudaGraphBufferRegistry`, `cuda_graph_config.prefill`, token-size buckets and padding, capture/replay lifecycle, eligibility checks, and eager fallback.
3. Keeping model runners responsible only for input preparation and (rarely) custom forward semantics.
4. Not building a parallel PCG implementation inside SGLang-Omni.

## The one upstream blocker to name explicitly

Main's `PrefillCudaGraphRunner.can_run_graph` currently rejects externally attached embeddings:

```python
if forward_batch.input_embeds is not None:
    return False
if forward_batch.replace_embeds is not None:
    return False
```

Since the initial scope below is exactly input-embeds-based prefill, **no first target can pass eligibility until this is lifted**. The lift is small, not a redesign: under `tc_piecewise`, the runner already registers a static `input_embeds` buffer slot for multimodal models (`registry.has_slot("input_embeds")`, `_input_embeds_arg_idx` plumbing into the captured layer stack). The missing piece is for `load_batch` / replay-prepare to copy `forward_batch.input_embeds` into that slot and for `can_run_graph` to accept the case when the slot exists. This should land as a small PR in SGLang main, not as an Omni-side patch. Note that BCG is text-only by design (it forces `input_embeds=None`), so the Omni embedding path effectively requires `--cuda-graph-backend-prefill tc_piecewise` until/unless BCG grows equivalent support.

## Proposed direction

Introduce one graph-aware prefill dispatch point in the shared Omni `ModelRunner`:

```python
# Model-specific input/state preparation. Mutates ForwardBatch only
# (e.g. attaches input_embeds); no forward execution, no attn metadata.
self.before_prefill(forward_batch, schedule_batch, requests)

# Explicit opt-out: a runner that truly owns its forward returns a result here
# (and accepts that it stays eager). Default returns None.
result = self.custom_prefill_forward(forward_batch, schedule_batch, requests)

if result is None:
    # Shared graph-aware path: upstream ModelRunner.forward decides
    # PCG replay vs eager via can_run_graph; Omni adds nothing here.
    result = self.tp_worker.forward_batch_generation(forward_batch)
```

The contract:

* `before_prefill` prepares or attaches model-specific tensor state to the `ForwardBatch`. It must not run the forward or touch attention metadata.
* `custom_prefill_forward` returning non-`None` is an explicit, observable opt-out of PCG — the escape hatch, not the default.
* Upstream owns graph buckets, static buffers, attention metadata initialization, capture, replay, and eligibility. Omni model runners manage none of it.

Deliberately **not** proposed: passing per-batch forward callables into the graph runner. `tc_piecewise` compiles and captures the model's forward at initialization; a callback-shaped API cannot ride that machinery and would drift into the parallel implementation this issue forbids. The dispatch contract is data-shaped (ForwardBatch mutation), with at most a small struct for observability:

```python
@dataclass
class PrefillGraphDecision:
    used_graph: bool
    fallback_reason: str | None = None  # e.g. "custom_forward", "input_embeds_unsupported", "over_max_tokens"
```

## Initial scope

Start with custom prefill paths whose dynamic behavior reduces to stable tensor inputs — models that replace token IDs with precomputed or projected `input_embeds`:

* Qwen3-TTS prefill (`models/qwen3_tts/model_runner.py`) — simplest: pure input-embeds prefill
* Qwen3-Omni talker prefill (`models/qwen3_omni/talker_model_runner.py`)
* MOSS-TTS / MOSS-TTS-Local prefill

Multimodal thinker paths stay eager initially: deepstack visual embeds and visual pos masks are extra forward arguments beyond `input_embeds`, the embed injection involves data-dependent Python, and the thinker additionally mutates module state for hidden capture (`layers_to_capture` forward hooks), which is unsafe under replay. They should still route through the same dispatch and report a fallback reason rather than keeping a separate architecture. (The thinker already returns `CaptureHiddenMode.NULL` from `requested_capture_hidden_mode_prefill` to avoid graph-mode mismatches — this proposal makes that contract explicit.)

## Compatibility and fallback

The generic path must fail closed to eager execution for any batch that is not graph-safe, including: dynamic Python control flow, unsupported custom ops, deepstack/multimodal shapes not yet supported, token counts above the largest captured bucket, hidden-capture modes differing from capture, forward-hook or module-state mutation, and upstream `can_run_graph` rejections. Fallback must expose a reason/counter so profiling can distinguish compile/warmup, capture, replay, and eager (with reason).

## Acceptance criteria

* [ ] Standard text-prefill behavior unchanged.
* [ ] At least one custom-input TTS prefill path (suggest Qwen3-TTS) runs through upstream PCG with no model-specific CUDA-graph code.
* [ ] The upstream `can_run_graph` input-embeds lift is merged in SGLang main (or an agreed interim plan exists tied to the version bump from 0.5.12.post1).
* [ ] Shared Omni `ModelRunner` owns graph-vs-eager dispatch for standard and custom prefill; hook ordering stays prepare/before → forward → post-prefill.
* [ ] Ineligible batches fall back to eager automatically, with observable reasons.
* [ ] No duplication of upstream capture buckets, buffers, attention metadata, or replay lifecycle in Omni.
* [ ] Unit tests: standard prefill; graph-eligible custom prefill; ineligible fallback; hook ordering; over-max-token fallback.
* [ ] Generated semantic tokens / codec codes match PCG-disabled execution under deterministic settings; no meaningful WER / speaker-similarity regression.
* [ ] Nsight (or equivalent) confirms reduced CPU launch gaps during eligible prefill.

## Benchmark matrix

For the first supported model, PCG on vs off (backend `tc_piecewise`), across: short/medium/long prefill token counts; single- and multi-request prefill batches; concurrency 1 and saturated; chunked-prefill boundaries; token counts above the largest bucket. Report: prefill wall time, TTFT / time-to-first-codec-frame, e2e latency, throughput, GPU busy %, inter-kernel gap distribution, startup compile+capture time, extra GPU memory, output parity, applicable quality metrics.

## Non-goals

* A second PCG implementation inside SGLang-Omni.
* Forcing multimodal/deepstack prefill through PCG before it is graph-safe.
* Replacing full CUDA graphs for fixed-shape autoregressive decode.
* Per-model one-off CUDA-graph classes (see #344 — that issue targets talker-specific CUDA-graph behavior; this one targets the shared prefill abstraction so compatible models benefit without one-off integrations).

## Related code

SGLang-Omni: `sglang_omni/model_runner/base.py` (`_prepare_and_forward`, hooks), `sglang_omni/model_runner/thinker_model_runner.py`, model-specific `custom_prefill_forward` implementations listed above.

SGLang main: `python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py` (`PrefillCudaGraphRunner`, `can_run_graph`, `input_embeds` slot plumbing), `cuda_graph_config.py` (per-phase backends, `tc_piecewise`), `cuda_graph_buffer_registry.py`, `input_buffers.py`. Pinned-release equivalents: `model_executor/piecewise_cuda_graph_runner.py` in `sglang==0.5.12.post1`.

**Comment by cr-gao** [comment]
Hi, I did #1161 (thinker multimodal embed merge; per-request device->host syncs 96->0), as those syncs
are part of what keeps thinker prefill out of a graph, so I went looking at what's left.

Two things from the pinned `0.5.12.post1`:

- The `input_embeds` rejection in `can_run` (`piecewise_cuda_graph_runner.py:460`) is a
  `TODO: fix it`, and the static `input_embeds` buffer it gates is already fully built in
  that same file (`:274-288` onward — alloc, capture, replay, padding). The gap looks like a
  contract mismatch: Omni attaches `forward_batch.input_embeds`, PCG wants the embeds copied
  into its pre-allocated buffer.
- Deepstack is the real gap. The thinker passes `input_deepstack_embeds` as a separate kwarg
  (`thinker_model_runner.py:355-361`) and `PrefillInputBuffers` has no slot for it. But deepstack
  is visual-only, so audio-only multimodal prefill needs no new buffer.

If you're not active on this, I'd take the shared `ModelRunner` graph/eager dispatch with the
thinker as the first model, leaving Qwen3-TTS enablement open, benchmarked on `audio -> text`,
which #1018 currently lists as unmeasured above c4. If you'd rather own the dispatch, happy to
build the thinker enablement on top of yours instead.

Either way I'd leave deepstack out of this pass as it needs a new upstream buffer and is already a
Non-goal here.

Not yet verified: whether `model_config.is_multimodal` is True for the thinker. If it isn't, the
static buffer is never allocated and this changes shape, spiking that this week.

**Comment by cr-gao** [comment]
Spiked it on H800 / Qwen3-Omni-30B-A3B-Instruct. Two things that change the framing:

- `can_run` is never reached. `custom_prefill_forward` calls `outer.model(...)` directly, bypassing
  `ModelRunner.forward` where dispatch lives — so the `input_embeds` rejection isn't what blocks the
  thinker. It takes over 4/4 fixtures including text-only, so all thinker prefill is eager today.
- Lifting that guard is the wrong direction anyway. Main's allowlist comments say embed-carrying
  batches run eager by design, and `general_mm_embed_routine` composes embeds inside the model
  forward and copies them into PCG's slot. `fun_asr` / `moss_transcribe_diarize` / `qwen3_asr`
  already conform; the thinker attaching externally is the outlier.

Since writing that I found #1035 and #1048 — @stephenkgli had already done per-model prefill PCG for
MOSS-TD along these lines and closed it himself, pending the 0.5.15 phase-aware lifecycle alignment
(#1183). That ordering is right; I'm not starting implementation ahead of it.

**Comment by AkazaAkane** [comment]
> Spiked it on H800 / Qwen3-Omni-30B-A3B-Instruct. Two things that change the framing:
> 
>     * `can_run` is never reached. `custom_prefill_forward` calls `outer.model(...)` directly, bypassing
>       `ModelRunner.forward` where dispatch lives — so the `input_embeds` rejection isn't what blocks the
>       thinker. It takes over 4/4 fixtures including text-only, so all thinker prefill is eager today.
> 
>     * Lifting that guard is the wrong direction anyway. Main's allowlist comments say embed-carrying
>       batches run eager by design, and `general_mm_embed_routine` composes embeds inside the model
>       forward and copies them into PCG's slot. `fun_asr` / `moss_transcribe_diarize` / `qwen3_asr`
>       already conform; the thinker attaching externally is the outlier.
> 
> 
> Since writing that I found [#1035](https://github.com/sgl-project/sglang-omni/pull/1035) and [#1048](https://github.com/sgl-project/sglang-omni/issues/1048) — [@stephenkgli](https://github.com/stephenkgli) had already done per-model prefill PCG for MOSS-TD along these lines and closed it himself, pending the 0.5.15 phase-aware lifecycle alignment ([#1183](https://github.com/sgl-project/sglang-omni/pull/1183)). That ordering is right; I'm not starting implementation ahead of it.

Thanks for contributing. We can start after the sglang bump merged.
